src/reflection.ts

```diff
@@ -839,11 +840,7 @@ export class Parser {
       throw new Error(`Malformed schema: union discriminator field is not a scalar`);
     }
 
-    const parseDiscriminator = this.readScalarLambda(
-      discriminatorField,
-      discriminatorfieldType.index(),
```

Comment:
This is the incorrect `typeIndex`. It should be the `typeIndex` from the call-site which is maybe the "parent" typeIndex?

Reply:
Yeah, `readScalarLambda` should take the `typeIndex` of the parent (yeah, that looks poorly documented). If you look at it, it uses that to figure out whether it should be treating the memory buffer as a table or a struct in order to actually locate the field in memory. The type index we had been using was an index into the enum list, which can both be out-of-range and is the wrong list.

In most use-cases, we were probably getting lucky because typically most schemas have more objects than enum types and most objects are tables not structs.

Writing a test will be slightly irritating, because it will depend on the order in which types appear in their respective lists of the reflection schema, and I don't know how consistent that is. It should be pretty easy to get to happen once (create a schema with some structs, some tables, and figure out how to reorder things so that the enum index would have sent you to a struct type); that test will end up relying on the reflection schema generation being relatively stable.